File: opensource/kernel/src/main/java/com/sos/scheduler/engine/eventbus/ColdEventBus.java
package com.sos.scheduler.engine.eventbus;

import com.sos.scheduler.engine.data.event.Event;
import com.sos.scheduler.engine.eventbus.annotated.ColdMethodEventSubscriptionFactory;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Collection;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;

import static com.google.common.base.Throwables.propagate;

public class ColdEventBus extends AbstractEventBus implements Runnable {
    private static final Logger logger = LoggerFactory.getLogger(ColdEventBus.class);

    private final LinkedBlockingQueue<Call> callQueue = new LinkedBlockingQueue<Call>();

    public ColdEventBus() {
        super(ColdMethodEventSubscriptionFactory.singleton);
    }

    @Override public final void publish(Event e) {
        publish(calls(e));
    }

    final void publish(Collection<Call> c) {
        callQueue.addAll(c);
    }

    public final void run() {
        try {
            dispatchEvents(true);
        } catch (InterruptedException ignored) {}
    }

    public final void dispatchEvents() {
        try {
            dispatchEvents(false);
        } catch (InterruptedException e) { throw propagate(e); }    // Kommt nicht vor
    }

    private void dispatchEvents(boolean wait) throws InterruptedException {
        while (true) {
            Call call = wait? callQueue.poll(Long.MAX_VALUE, TimeUnit.DAYS) : callQueue.poll();
            if (call == null) break;
            logger.trace("dispatch "+call);
            dispatchCall(call);
        }
    }
}

File: engine-kernel/src/main/scala/com/sos/scheduler/engine/eventbus/ColdEventBus.scala
package com.sos.scheduler.engine.eventbus

import com.sos.jobscheduler.common.scalautil.Logger
import com.sos.jobscheduler.data.event.{Event, KeyedEvent}
import com.sos.scheduler.engine.eventbus.ColdEventBus._
import com.sos.scheduler.engine.eventbus.annotated.ColdMethodEventSubscriptionFactory
import java.util.concurrent.{LinkedBlockingQueue, TimeUnit}
import scala.collection.JavaConversions._

final class ColdEventBus extends AbstractEventBus with Runnable {

  protected def factory = ColdMethodEventSubscriptionFactory.singleton

  private val callQueue = new LinkedBlockingQueue[Call]

  def publish(e: KeyedEvent[Event]) {
    publish(calls(e))
  }

  private[eventbus] def publish(calls: Iterable[Call]) =
    callQueue.addAll(calls)

  def run() = dispatchEvents(true)

  def dispatchEvents(): Unit = dispatchEvents(false)

  private def dispatchEvents(wait: Boolean) {
    while (true) {
      val call = poll(wait)
      if (call == null)
        return
      logger.trace(s"dispatch $call")
      dispatchCall(call)
    }
  }

  private def poll(wait: Boolean) =
    if (wait)
      callQueue.poll(Long.MaxValue, TimeUnit.DAYS)
    else
      callQueue.poll()
}

object ColdEventBus {
  private val logger = Logger(getClass)
}
